### PR TITLE
Bigtable: Create Column Family snippets

### PR DESCRIPTION
**Pull request by sangramql**

### CODE REVIEW

**Comment by sangramql** [comment]
@tseaver , @AVaksman, @sduskis, Please review the snippets for column family.

File: bigtable/docs/snippets.py [hunk 1]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+    client = Client(admin=True)
+    instance = client.instance(INSTANCE_ID)
+    table = instance.table(TABLE_ID)
+
+    # Define the GC policy to retain only the most recent 2 versions
+    max_versions_rule = column_family.MaxVersionsGCRule(2)
+
+    column_family_obj = table.column_family('cf1', max_versions_rule)
+    column_family_obj.create()
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: For the example purposes we only need to show how to create a `max_versions_rule` object. 
As such, we do not need to import `Client` and we do not need `client`, `instance` and `table` variables above.

File: bigtable/docs/snippets.py [hunk 2]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+    # [END bigtable_create_MaxVersionsGCRule]
+    rule = str(column_family_obj.to_pb())
+    assert "max_num_versions: 2" in rule
+    column_family_obj.delete()
+
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: To test the object you could use
```python
from google.cloud.bigtable.column_family import MaxVersionsGCRule
assert isinstance(max_versions_rule, MaxVersionsGCRule)
assert max_versions_rule.max_num_versions == 2
```

File: bigtable/docs/snippets.py [hunk 3]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+def test_bigtable_create_MaxAgeGCRule():
+    # [START bigtable_create_MaxAgeGCRule]
+    from google.cloud.bigtable import Client
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Please remove this import as per above.

File: bigtable/docs/snippets.py [hunk 4]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    from google.cloud.bigtable import column_family
+
+    client = Client(admin=True)
+    instance = client.instance(INSTANCE_ID)
+    table = instance.table(TABLE_ID)
+
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Please remove these three lines as per above.

File: bigtable/docs/snippets.py [hunk 5]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    # Define the GC rule to retain data with max age of 5 days
+    max_age_rule = column_family.MaxAgeGCRule(datetime.timedelta(days=5))
+
+    column_family_obj = table.column_family('cf1', max_age_rule)
+    column_family_obj.create()
+
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Remove these two lines as per above as well.

File: bigtable/docs/snippets.py [hunk 6]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    # [END bigtable_create_MaxAgeGCRule]
+    rule = str(column_family_obj.to_pb())
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: To test you could use
```python
from google.cloud.bigtable.column_family import MaxAgeGCRule
assert isinstance(max_age_rule, MaxAgeGCRule)
assert max_age_rule.max_age.days == 5
```

File: bigtable/docs/snippets.py [hunk 7]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    assert "max_age" in rule
+    assert "seconds: 432000" in rule
+    column_family_obj.delete()
+
+
+def test_bigtable_create_GCRuleUnion():
+    # [START bigtable_create_GCRuleIntersection]
+    from google.cloud.bigtable import Client
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Please remove this import as per above

File: bigtable/docs/snippets.py [hunk 8]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    from google.cloud.bigtable import column_family
+
+    client = Client(admin=True)
+    instance = client.instance(INSTANCE_ID)
+    table = instance.table(TABLE_ID)
+
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Please remove these three lines as per above

File: bigtable/docs/snippets.py [hunk 9]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    max_versions_rule = column_family.MaxVersionsGCRule(2)
+    max_age_rule = column_family.MaxAgeGCRule(datetime.timedelta(days=5))
+
+    union_rule = column_family.GCRuleUnion([max_versions_rule, max_age_rule])
+
+    column_family_obj = table.column_family('cf1', union_rule)
+    column_family_obj.create()
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Remove these two lines as per above please

File: bigtable/docs/snippets.py [hunk 10]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+    # [END bigtable_create_GCRuleIntersection]
+    rule = str(column_family_obj.to_pb())
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: To test you could use 
```python
from google.cloud.bigtable.column_family import GCRuleUnion
assert isinstance(union_rule, GCRuleUnion)
assert max_versions_rule in union_rule.rules
assert max_age_rule in union_rule.rules
```

File: bigtable/docs/snippets.py [hunk 11]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    assert "union" in rule
+    assert "max_age" in rule
+    assert "seconds: 432000" in rule
+    assert "max_num_versions: 2" in rule
+    column_family_obj.delete()
+
+
+def test_bigtable_create_GCRuleIntersection():
+    # [START bigtable_create_GCRuleUnion]
+    from google.cloud.bigtable import Client
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Please remove this import as per above as well

File: bigtable/docs/snippets.py [hunk 12]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    from google.cloud.bigtable import column_family
+
+    client = Client(admin=True)
+    instance = client.instance(INSTANCE_ID)
+    table = instance.table(TABLE_ID)
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Please remove these three lines as per above as well

File: bigtable/docs/snippets.py [hunk 13]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    max_versions_rule = column_family.MaxVersionsGCRule(2)
+    max_age_rule = column_family.MaxAgeGCRule(datetime.timedelta(days=5))
+
+    intersection_rule = column_family.GCRuleIntersection([
+        max_versions_rule,
+        max_age_rule
+        ])
+
+    column_family_obj = table.column_family('cf1', intersection_rule)
+    column_family_obj.create()
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Please remove these two lines as per above as well.

File: bigtable/docs/snippets.py [hunk 14]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+    # [END bigtable_create_GCRuleUnion]
+
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: To test you could use 
```python
from google.cloud.bigtable.column_family import GCRuleIntersection
assert isinstance(intersection_rule, GCRuleIntersection)
assert max_versions_rule in intersection_rule.rules
assert max_age_rule in intersection_rule.rules
```

File: bigtable/docs/snippets.py [hunk 15]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    column_family_id = "column_family_id1"
+    column_family_obj = table.column_family(column_family_id)
+    column_family_obj.delete()
+    # [END bigtable_delete_column_family]
+    column_families = table.list_column_families()
+    assert column_family_id not in column_families
+
+
+def test_bigtable_create_MaxVersionsGCRule():
+    # [START bigtable_create_MaxVersionsGCRule]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: I could suggest renaming target to just 
`# [START bigtable_MaxVersionsGCRule]` or `# [START bigtable_MaxVersionsGCRule_object]`
for clarification.

File: bigtable/docs/snippets.py [hunk 16]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    instance = client.instance(INSTANCE_ID)
+    table = instance.table(TABLE_ID)
+
+    # Define the GC policy to retain only the most recent 2 versions
+    max_versions_rule = column_family.MaxVersionsGCRule(2)
+
+    column_family_obj = table.column_family('cf1', max_versions_rule)
+    column_family_obj.create()
+
+    # [END bigtable_create_MaxVersionsGCRule]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Follow the tag format above. (if different format is chosen)

File: bigtable/docs/snippets.py [hunk 17]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    rule = str(column_family_obj.to_pb())
+    assert "max_num_versions: 2" in rule
+    column_family_obj.delete()
+
+
+def test_bigtable_create_MaxAgeGCRule():
+    # [START bigtable_create_MaxAgeGCRule]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Follow the tag format above. (if different format is chosen)

File: bigtable/docs/snippets.py [hunk 18]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    instance = client.instance(INSTANCE_ID)
+    table = instance.table(TABLE_ID)
+
+    # Define the GC rule to retain data with max age of 5 days
+    max_age_rule = column_family.MaxAgeGCRule(datetime.timedelta(days=5))
+
+    column_family_obj = table.column_family('cf1', max_age_rule)
+    column_family_obj.create()
+
+    # [END bigtable_create_MaxAgeGCRule]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: Follow the tag format above. (if different format is chosen)

File: bigtable/docs/snippets.py [hunk 19]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+    max_versions_rule = column_family.MaxVersionsGCRule(2)
+    max_age_rule = column_family.MaxAgeGCRule(datetime.timedelta(days=5))
+
+    union_rule = column_family.GCRuleUnion([max_versions_rule, max_age_rule])
+
+    column_family_obj = table.column_family('cf1', union_rule)
+    column_family_obj.create()
+
+    # [END bigtable_create_GCRuleIntersection]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: should be `...GCRuleUnion` tag
Follow the tag format above. (if different format is chosen)

File: bigtable/docs/snippets.py [hunk 20]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+    # [END bigtable_create_MaxAgeGCRule]
+    rule = str(column_family_obj.to_pb())
+    assert "max_age" in rule
+    assert "seconds: 432000" in rule
+    column_family_obj.delete()
+
+
+def test_bigtable_create_GCRuleUnion():
+    # [START bigtable_create_GCRuleIntersection]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: should be `...GCRuleUnion` tag
Follow the tag format above. (if different format is chosen)

File: bigtable/docs/snippets.py [hunk 21]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+    rule = str(column_family_obj.to_pb())
+    assert "union" in rule
+    assert "max_age" in rule
+    assert "seconds: 432000" in rule
+    assert "max_num_versions: 2" in rule
+    column_family_obj.delete()
+
+
+def test_bigtable_create_GCRuleIntersection():
+    # [START bigtable_create_GCRuleUnion]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: should be `...GCRuleIntersection` tag
Follow the tag format above. (if different format is chosen)

File: bigtable/docs/snippets.py [hunk 22]
```diff
@@ -160,6 +170,183 @@ def test_bigtable_create_app_profile():
...
+
+    intersection_rule = column_family.GCRuleIntersection([
+        max_versions_rule,
+        max_age_rule
+        ])
+
+    column_family_obj = table.column_family('cf1', intersection_rule)
+    column_family_obj.create()
+
+    # [END bigtable_create_GCRuleUnion]
```
Comment (AVaksman) [inline bigtable/docs/snippets.py]: should be `...GCRuleIntersection` tag
Follow the tag format above. (if different format is chosen)

File: bigtable/google/cloud/bigtable/column_family.py [hunk 23]
```diff
@@ -43,6 +43,12 @@ class GarbageCollectionRule(object):
 class MaxVersionsGCRule(GarbageCollectionRule):
     """Garbage collection limiting the number of versions of a cell.
 
+    For example:
+
+    .. literalinclude:: snippets.py
+        :start-after: [START bigtable_create_MaxVersionsGCRule]
+        :end-before: [END bigtable_create_MaxVersionsGCRule]
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/column_family.py]: Update tags if format changed in snippets.py

File: bigtable/google/cloud/bigtable/column_family.py [hunk 24]
```diff
@@ -70,6 +76,12 @@ def to_pb(self):
 class MaxAgeGCRule(GarbageCollectionRule):
     """Garbage collection limiting the age of a cell.
 
+    For example:
+
+    .. literalinclude:: snippets.py
+        :start-after: [START bigtable_create_MaxAgeGCRule]
+        :end-before: [END bigtable_create_MaxAgeGCRule]
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/column_family.py]: Update tags if format changed in snippets.py

File: bigtable/google/cloud/bigtable/column_family.py [hunk 25]
```diff
@@ -98,6 +110,12 @@ def to_pb(self):
 class GCRuleUnion(GarbageCollectionRule):
     """Union of garbage collection rules.
 
+    For example:
+
+    .. literalinclude:: snippets.py
+        :start-after: [START bigtable_create_GCRuleUnion]
+        :end-before: [END bigtable_create_GCRuleUnion]
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/column_family.py]: Update tags if format changed in snippets.py

File: bigtable/google/cloud/bigtable/column_family.py [hunk 26]
```diff
@@ -127,6 +145,12 @@ def to_pb(self):
 class GCRuleIntersection(GarbageCollectionRule):
     """Intersection of garbage collection rules.
 
+    For example:
+
+    .. literalinclude:: snippets.py
+        :start-after: [START bigtable_create_GCRuleIntersection]
+        :end-before: [END bigtable_create_GCRuleIntersection]
```
Comment (AVaksman) [inline bigtable/google/cloud/bigtable/column_family.py]: Update tags if format changed in snippets.py

**Comment by AVaksman** [comment]
I am getting an error on `column_family_obj.create()`
`google.api_core.exceptions.FailedPrecondition: 400 Could not modify schema of projects/grass-clump-479/instances/snippet--1542650087526/tables/tabl-1--1542650087526due to concurrent modifications. This could mean another change to this table's schema, or a change to the replication topology (adding/removing a cluster).`

I am suggesting to run admin snippet tests and data snippet tests on separate instances to avoid any trouble with _replication_.
This might be cleaner going forward.

**Comment by sangramql** [comment]
I that's the case then it looks ideal for creating separate snippets files for instance admin snippets and table admin snippets. 

In which case, there would be separate setup and teardown. which would handle new instance for separate files.

**Comment by sangramql** [comment]
I'll work on this PR once 6484 is merged, as we can follow same logic of either single file multiple files snippets.
Though I'll locally fix any issues/reviews from this PR but won't submit till 6484 is merged.

**Comment by sangramql** [comment]
Closing this as new PR  #7014  created for column family.

**Pull request closed without merging**